RxCocoa/Common/DelegateProxyType.swift

```diff
@@ -210,6 +213,14 @@ extension DelegateProxyType {
             proxy.setForwardToDelegate(nil, retainDelegate: retainDelegate)
         }
     }
+    
+    public static func extend<Object: AnyObject>(with factory: @escaping ((Object) -> AnyObject)) {
+        factories.append({ ($0 as? Object).map(factory) })
+    }
+    
+    public static func createProxyForObject(_ object: AnyObject) -> AnyObject {
+        return factories.reversed().reduce(AnyObject?.none) { $0 ?? $1(object) }!
```

Comment:
Hi @tarunon ,

I think this implementation makes an assumption that `extend` function callers perform calls in correct order.

This seems fragile to me and it's not exactly clear from the public interface there is this requirement. I would expect that implementation would resolve class hierarchies correctly.

So this should also work:
```
RxScrollViewDelegateProxy.extend { (parentObject: UITableViewSubclassSubclass1) in
         ExtendTableViewDelegateProxy(parentObject: parentObject)
}
RxScrollViewDelegateProxy.extend { (parentObject: UITableViewSubclass1) in
         ExtendTableViewDelegateProxy(parentObject: parentObject)
}
```

Since you are also saying 
> I think I will separate type around factory from DelegateProxy when I implement DelegateProxy with generics type. Because we cannot make static property using generics type.

Let's then maybe change this storage right now and move everything into one private static dictionary somewhere.
```
static var proxyMakers: [AnyClass: (AnyObject) -> AnyObject] // or some more idiomatic name
```

I think the resolution algorithm should look something like:
```
public static func makeProxy(for: AnyObject) -> AnyObject {
     MainScheduler.ensureExecutingOnScheduler(some meaningful message)
     var type = type(of: for)
     while type != nil {
          defer {
              type = type.parent
          }
          if let result = proxyMakers[type].map { $0(for) } {
               return result
          }

     }

     fatalError("Some meaningful error that describes we couldn't resolve delegate proxy for "\(for). Please use xxx to register delegate proxy")
} 
```

I don't think this should cause any performance concerns.